src/Icicle/Source/Checker/Constraint.hs

```diff
@@ -93,21 +103,25 @@ constraintsQ env q
 -- The constraints flow upwards, bottom-up, but are also discharged and simplified at every level.
 -- The substitutions from discharged constraints also flow upwards.
 --
--- The environment mapping flows downwards, with new names only available in lexically nested subexpressions and so on.
+-- The environment mapping flows downwards, with new names only available in lexically
+-- nested subexpressions and so on.
 --
--- Shoehorning these two differently flowing contexts into a state monad was a mistake, and is a constant source of bugs.
---
-generateQ :: Ord n => Query a n -> Gen a n (Query'C a n, SubstT n)
+generateQ
+  :: Ord n
+  => Query a n
+  -> CheckContext a n
+  -> Gen a n (Query'C a n, SubstT n, CheckContext a n)
```

Comment:
we want something like this type for all of the places where the environment/context was used before:
```
=> Query a n
-> Map n FunctionType
-> Gen a n (Query'C a n, SubstT n, [Constraint a n])
```
because the Environment only goes down lexically, and the constraints and type substitutions just have to bubble up